admin/view_log.php

```diff
@@ -0,0 +1,55 @@
+<?php
+/**
+ * @package admin
+ * @copyright Copyright 2016 Zen Cart Development Team
+ * @license http://www.zen-cart.com/license/2_0.txt GNU Public License V2.0
+ * @version $Id: $
+ */
+
+  require('includes/application_top.php');
+  require('includes/admin_html_head.php');
+?>
+</head>
+<body>
+<!-- header //-->
+<div class="header-area">
+<?php require(DIR_WS_INCLUDES . 'header.php'); ?>
+</div>
+<!-- header_eof //-->
+
+<!-- body //-->
+<div class="pageHeading"><?php echo HEADING_TITLE; ?></div>
+<br />
+<?php 
+  if (!defined('DIR_FS_LOGS')) define('DIR_FS_LOGS', DIR_FS_CATALOG . 'logs');
+   $pathname = DIR_FS_LOGS . "/";
+   $filename = $_GET['logname']; 
+   $file = $pathname . $filename;
+   $rpdir = realpath($file);
+   $rppathdir = realpath($pathname);
+   if ( (substr($rpdir, 0, strlen($rppathdir)) != $rppathdir) || 
+        (strlen($rpdir) < strlen($rppathdir)) ) { 
+      // hacking logname parameter in URL
+      echo FILE_NOT_FOUND; 
+      zen_record_admin_activity("Possible hacking in log viewer - looking at " . $rppathdir);
+   } else if (file_exists($file)) {
+      echo '<strong>' . PATH . '</strong>' .  $pathname . '<br />'; 
+      echo '<strong>' . FILENAME . '</strong>' .  $filename . '<br /><br />'; 
+      $file_array = @file($file);
+      $file_contents = @implode('', $file_array);
+      echo nl2br(zen_output_string_protected($file_contents)); 
+   } else { 
+      echo FILE_NOT_FOUND; 

```

Comment:
The first one really isn't a file not found, it's a "you don't have
permission," but I don't want to say that - or do you think I should?
The first one occurs when they fiddle with the URL to look at a file
outside of the logs directory.

On Thu, Jun 2, 2016 at 9:49 PM, Chris Brown <notifications@github.com>
wrote:

> In admin/view_log.php
> <https://github.com/zencart/zencart/pull/950#discussion_r65647056>:
>
> > +   $file = $pathname . $filename;
> > +   $rpdir = realpath($file);
> > +   $rppathdir = realpath($pathname);
> > +   if ( (substr($rpdir, 0, strlen($rppathdir)) != $rppathdir) ||
> > +        (strlen($rpdir) < strlen($rppathdir)) ) {
> > +      // hacking logname parameter in URL
> > +      echo FILE_NOT_FOUND;
> > +      zen_record_admin_activity("Possible hacking in log viewer - looking at " . $rppathdir);
> > +   } else if (file_exists($file)) {
> > +      echo '<strong>' . PATH . '</strong>' .  $pathname . '<br />';
> > +      echo '<strong>' . FILENAME . '</strong>' .  $filename . '<br /><br />';
> > +      $file_array = @file($file);
> > +      $file_contents = @implode('', $file_array);
> > +      echo nl2br(zen_output_string_protected($file_contents));
> > +   } else {
> > +      echo FILE_NOT_FOUND;
>
> Thoughts on simplifying the if...elseif...else?
> I'm wordering why there are 2 different paths to an identical
> FILE_NOT_FOUND result.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zencart/zencart/pull/950/files/2517190126e2709dca526dff5a4f6e6a14472fac#r65647056>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEMC1nfF-fxjDGo6yG3kaNayCAs3ZdsAks5qH4gYgaJpZM4IpRh6>
> .
>